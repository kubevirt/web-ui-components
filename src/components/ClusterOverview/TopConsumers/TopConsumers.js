import React from 'react';
import { UtilizationBar } from 'patternfly-react';

import PropTypes from 'prop-types';

import { getFormElement } from '../../Form';
import { DROPDOWN } from '../../Form/constants';
import { PodModel, VirtualMachineModel } from '../../../models';
import {
  DashboardCard,
  DashboardCardBody,
  DashboardCardHeader,
  DashboardCardTitle,
  DashboardCardTitleHelp,
} from '../../Dashboard/DashboardCard';

const sortConsumers = (metrics, filter, sortBy) => {
  const metricKey = Object.keys(metrics).find(key => metrics[key].title === sortBy);
  const filteredConsumers =
    filter !== 'All' ? metrics[metricKey].consumers.filter(c => c.kind === filter) : metrics[metricKey].consumers;
  const max = Math.max(...filteredConsumers.map(c => c.usage));
  return filteredConsumers
    .map(c => ({
      now: Math.round((100 * c.usage) / max),
      description: c.name,
    }))
    .sort((a, b) => b.now - a.now);
};

class TopConsumersBody extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      filter: 'All',
      sortBy: 'CPU',
    };
  }

  onDropdownChange = (key, newValue) => this.setState({ [key]: newValue });

  render() {
    const sortByDropdown = {
      id: 'sort-by-dropdown',
      type: DROPDOWN,
      choices: ['CPU', 'Memory', 'Network', 'Storage'],
      onChange: newValue => this.onDropdownChange('sortBy', newValue),
      value: this.state.sortBy,
    };
    const filterDropdown = {
      id: 'filter-dropdown',
      type: DROPDOWN,
      choices: ['All', PodModel.kind, VirtualMachineModel.kind],
      onChange: newValue => this.onDropdownChange('filter', newValue),
      value: this.state.filter,
    };
    return (
      <div>
        <div className="kubevirt-consumers__filters">
          {getFormElement(filterDropdown)}
          {getFormElement(sortByDropdown)}
        </div>
        {sortConsumers(this.props.metrics, this.state.filter, this.state.sortBy).map((c, index) => (
          <UtilizationBar key={index} {...c} className="kubevirt-consumers__bar" />
        ))}
      </div>
    );
  }
}

TopConsumersBody.propTypes = {
  metrics: PropTypes.object.isRequired,
};

const TopConsumers = ({ metrics, loaded }) => (
  <DashboardCard>
    <DashboardCardHeader>
      <DashboardCardTitle>Top Consumers</DashboardCardTitle>
      <DashboardCardTitleHelp>help for top consumers</DashboardCardTitleHelp>
    </DashboardCardHeader>
    <DashboardCardBody isLoading={!loaded}>
      <TopConsumersBody metrics={metrics} />
    </DashboardCardBody>
  </DashboardCard>
);

TopConsumers.defaultProps = {
  loaded: false,
};

TopConsumers.propTypes = {
  metrics: PropTypes.object.isRequired,
  loaded: PropTypes.bool,
};

export default TopConsumers;
