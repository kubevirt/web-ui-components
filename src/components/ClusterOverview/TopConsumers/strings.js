export const BY_CPU = 'By CPU';
export const CPU_DESC = 'CPU Time';

export const BY_MEMORY = 'By Memory';
export const MEMORY_DESC = 'Memory consumption';

export const BY_STORAGE = 'By Storage';
export const STORAGE_DESC = 'Filesystem IO Time';

export const BY_NETWORK = 'By Network';
export const NETWORK_DESC = 'Network consumption';

export const PODS_AND_VMS = 'Pods & VMs';
export const NODES = 'Nodes';

export const WORKLOADS = 'Workloads';
export const INFRASTRUCTURE = 'Infrastructure';
