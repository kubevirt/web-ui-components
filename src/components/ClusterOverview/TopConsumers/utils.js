import { formatBytes } from '../../../utils';
import { getConsumers } from '../../../selectors';

export const formatCpu = (cpuTime, fixed = 0) => Number(cpuTime).toFixed(fixed);

export const formatBytesWithUnits = bytes => {
  const formattedBytes = formatBytes(bytes);
  return `${formattedBytes.value} ${formattedBytes.unit}`;
};

export const getMetric = (results, metricName, format) => ({
  isLoading: !results,
  consumers: getConsumers(results, metricName, format),
});
