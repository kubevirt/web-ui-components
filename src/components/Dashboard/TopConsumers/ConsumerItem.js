import React from 'react';
import PropTypes from 'prop-types';
import { Row, Col, OverlayTrigger, ProgressBar, Tooltip } from 'patternfly-react';

export const ConsumerItem = ({ now, description, label }) => (
  <Row className="kubevirt-consumers__bar">
    <Col lg={4} md={4} sm={4} xs={4}>
      <OverlayTrigger
        overlay={<Tooltip id={`tooltip-for-${description}`}>{description}</Tooltip>}
        placement="top"
        trigger={['hover', 'focus']}
        rootClose={false}
      >
        <div className="kubevirt-consumers__bar-description">{description}</div>
      </OverlayTrigger>
    </Col>
    <Col lg={5} md={5} sm={5} xs={5} className="kubevirt-consumers__bar-column-progress">
      <div className="progress kubevirt-consumers__bar-progress">
        <ProgressBar min={0} max={100} now={now} key={1} isChild />
        <ProgressBar min={0} max={100} now={100 - now} key={2} bsClass="progress-bar progress-bar-remaining" isChild />
      </div>
    </Col>
    <Col lg={3} md={3} sm={3} xs={3} className="kubevirt-consumers__bar-label">
      <OverlayTrigger
        overlay={<Tooltip id={`tooltip-for-${label}`}>{label}</Tooltip>}
        placement="top"
        trigger={['hover', 'focus']}
        rootClose={false}
      >
        <span>{label}</span>
      </OverlayTrigger>
    </Col>
  </Row>
);

ConsumerItem.propTypes = {
  now: PropTypes.number.isRequired,
  description: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
};
