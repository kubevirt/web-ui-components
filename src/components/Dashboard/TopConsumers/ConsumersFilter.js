import React from 'react';
import PropTypes from 'prop-types';

export const ConsumersFilter = ({ children }) => <div className="kubevirt-consumers__filters">{children}</div>;

ConsumersFilter.propTypes = {
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
};
