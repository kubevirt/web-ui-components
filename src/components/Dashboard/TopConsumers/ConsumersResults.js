import React from 'react';
import { Row, Col } from 'patternfly-react';
import PropTypes from 'prop-types';

import { InlineLoading } from '../../Loading';

export const ConsumersResults = ({ type, description, children, isLoading, LoadingComponent }) => {
  let results = 'Not available';
  if (isLoading) {
    results = <LoadingComponent />;
  } else if (children) {
    results = children;
  }

  return (
    <div className="kubevirt-consumers__results">
      {type && description && (
        <Row className="kubevirt-consumers__description">
          <Col lg={4} md={4} sm={4} xs={4} className="kubevirt-consumers__metric-type">
            {type}
          </Col>
          <Col lg={8} md={8} sm={8} xs={8} className="kubevirt-consumers__metric-description">
            {description}
          </Col>
        </Row>
      )}
      {results}
    </div>
  );
};

ConsumersResults.defaultProps = {
  type: null,
  description: null,
  isLoading: false,
  LoadingComponent: InlineLoading,
  children: null,
};

ConsumersResults.propTypes = {
  type: PropTypes.string,
  description: PropTypes.string,
  isLoading: PropTypes.bool,
  LoadingComponent: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
};
