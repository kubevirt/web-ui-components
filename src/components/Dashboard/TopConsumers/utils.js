export const getMetricConsumers = ({ consumers }) => {
  const max = Math.max(...consumers.map(c => c.usage));
  return consumers
    .map(c => ({
      now: Math.round((100 * c.usage) / max),
      description: c.name,
      label: c.label,
    }))
    .sort((a, b) => b.now - a.now);
};
