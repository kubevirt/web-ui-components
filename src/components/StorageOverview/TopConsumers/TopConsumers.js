import React from 'react';
import PropTypes from 'prop-types';
import { get } from 'lodash';

import { Row, Col } from 'patternfly-react';
import {
  ChartGroup,
  ChartLine,
  ChartTheme,
  ChartTooltip,
  Chart,
  ChartAxis,
  ChartLegend,
  ChartVoronoiContainer,
} from '@patternfly/react-charts';

import { InlineLoading } from '../../Loading';
import { formatToShortTime } from '../../../utils';

import {
  DashboardCard,
  DashboardCardBody,
  DashboardCardHeader,
  DashboardCardTitle,
} from '../../Dashboard/DashboardCard';
import { StorageOverviewContext } from '../StorageOverviewContext/StorageOverviewContext';
import { getTopConsumerVectorStats } from '../../../selectors/prometheus/storage';

const TopConsumersBody = ({ topConsumers }) => {
  let results = 'No data available';
  const topConsumerStats = get(topConsumers, 'data.result', []);

  if (topConsumerStats.length) {
    const stats = getTopConsumerVectorStats(topConsumerStats);
    const { chartData, legends, maxCapacity, unit } = stats;
    const yTickValues = [
      0,
      Number((maxCapacity / 4).toFixed(1)),
      Number((maxCapacity / 2).toFixed(1)),
      Number(((3 * maxCapacity) / 4).toFixed(1)),
      maxCapacity,
      Number(((5 * maxCapacity) / 4).toFixed(1)),
      Number(((6 * maxCapacity) / 4).toFixed(1)),
    ];

    const chartLineList = chartData.map((data, i) => <ChartLine key={i} data={data} x={0} y={1} />);
    results = (
      <React.Fragment>
        <Row>
          <Col className="kubevirt-top-consumer__time-duration">Last 10 minutes</Col>
        </Row>
        <Row>
          <div>
            <Chart
              domain={{ y: [0, 1.5 * maxCapacity] }}
              theme={ChartTheme.light.multi}
              height={175}
              padding={{ top: 20, bottom: 20, left: 40, right: 17 }}
              containerComponent={
                <ChartVoronoiContainer
                  labels={datum => `${datum[1]} ${unit}`}
                  labelComponent={<ChartTooltip style={{ fontSize: 8, padding: 5 }} />}
                />
              }
              scale={{ x: 'time' }}
            >
              <ChartGroup>{chartLineList}</ChartGroup>
              <ChartAxis tickFormat={x => formatToShortTime(x)} style={{ tickLabels: { fontSize: 8, padding: 5 } }} />
              <ChartAxis
                label={`Requested capacity(${unit})`}
                dependentAxis
                tickValues={yTickValues}
                style={{ tickLabels: { fontSize: 8, padding: 5 }, axisLabel: { fontSize: 8, padding: 25 } }}
              />
            </Chart>
          </div>
          <ChartLegend
            data={legends}
            orientation="horizontal"
            gutter={10}
            height={30}
            padding={{ top: 5, bottom: 5, left: 20 }}
            theme={ChartTheme.light.multi}
            x={50}
            y={8}
          />
        </Row>
      </React.Fragment>
    );
  }

  return <div>{results}</div>;
};

export const TopConsumers = ({ topConsumers, LoadingComponent }) => (
  <DashboardCard>
    <DashboardCardHeader>
      <DashboardCardTitle>Top Projects by Requested Capacity</DashboardCardTitle>
    </DashboardCardHeader>
    <DashboardCardBody isLoading={!topConsumers} LoadingComponent={LoadingComponent}>
      <TopConsumersBody topConsumers={topConsumers} />
    </DashboardCardBody>
  </DashboardCard>
);

TopConsumersBody.defaultProps = {
  topConsumers: null,
};

TopConsumersBody.propTypes = {
  topConsumers: PropTypes.object,
};

TopConsumers.defaultProps = {
  topConsumers: null,
  LoadingComponent: InlineLoading,
};

TopConsumers.propTypes = {
  topConsumers: PropTypes.object,
  LoadingComponent: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
};

const TopConsumersConnected = () => (
  <StorageOverviewContext.Consumer>{props => <TopConsumers {...props} />}</StorageOverviewContext.Consumer>
);

export default TopConsumersConnected;
