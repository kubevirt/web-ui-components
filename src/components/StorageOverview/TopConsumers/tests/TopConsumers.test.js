import React from 'react';
import { shallow } from 'enzyme';

import { TopConsumers } from '../TopConsumers';
import { topConsumers, emptyTopConsumers } from '../fixtures/TopConsumers.fixture';

describe('<TopConsumers />', () => {
  it('no-data-renders-correctly', () => {
    const component = shallow(<TopConsumers topConsumers={emptyTopConsumers} />);
    expect(component).toMatchSnapshot();
  });
  it('data-renders-correctly', () => {
    const component = shallow(<TopConsumers topConsumers={topConsumers} />);
    expect(component).toMatchSnapshot({});
  });
  it('render-correctly-on-loading', () => {
    const component = shallow(<TopConsumers />);
    expect(component).toMatchSnapshot();
  });
});
