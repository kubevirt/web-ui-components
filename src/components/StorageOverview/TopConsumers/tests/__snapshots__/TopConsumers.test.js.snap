// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopConsumers /> data-renders-correctly 1`] = `Object {}`;

exports[`<TopConsumers /> no-data-renders-correctly 1`] = `
<DashboardCard
  className={null}
>
  <DashboardCardHeader
    className={null}
  >
    <DashboardCardTitle
      className={null}
    >
      Top Projects by Requested Capacity
    </DashboardCardTitle>
  </DashboardCardHeader>
  <DashboardCardBody
    LoadingComponent={[Function]}
    isLoading={false}
  >
    <TopConsumersBody
      topConsumers={
        Object {
          "data": Object {
            "result": Array [],
          },
        }
      }
    />
  </DashboardCardBody>
</DashboardCard>
`;

exports[`<TopConsumers /> render-correctly-on-loading 1`] = `
<DashboardCard
  className={null}
>
  <DashboardCardHeader
    className={null}
  >
    <DashboardCardTitle
      className={null}
    >
      Top Projects by Requested Capacity
    </DashboardCardTitle>
  </DashboardCardHeader>
  <DashboardCardBody
    LoadingComponent={[Function]}
    isLoading={true}
  >
    <TopConsumersBody
      topConsumers={null}
    />
  </DashboardCardBody>
</DashboardCard>
`;
